target/linux/ath79/image/mikrotik.mk

```diff
@@ -1,5 +1,16 @@
 include ./common-mikrotik.mk
 
+define Device/mikrotik_routerboard-493g
+  $(Device/mikrotik)
+  SOC := ar7161
+  DEVICE_MODEL := RouterBOARD 493G
+  IMAGE/sysupgrade.bin = append-kernel | kernel2minor -s 2048 -e -c | \
+	sysupgrade-tar kernel=$$$$@ | append-metadata
```

Comment:
You're not using lzma-loader here

Reply:
Am I not picking it up from [here](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/image/common-mikrotik.mk)?

N.B. I'm tftp booting this board with the initramfs image, and it's failing.

Reply:
Indeed, my bad. I'm running out of ideas. Do you have serial to see what's going on?

Reply:
I do, and I can see the bootloader pulls the image over the network, and then I see the lzma-loader kick in and decompress it, then ... nothing!

I have a theory something is being clobbered in ram (perhaps similar to [this change](https://github.com/openwrt/openwrt/commit/071a6372762d289b2fa5a8d1a138eff944c93c9e)).

```
RouterBOOT booter 6.41

RouterBoard 493G

CPU frequency: 680 MHz
  Memory size: 256 MiB
    NAND size: 128 MiB

Press any key within 2 seconds to enter setup..
trying bootp protocol..... OK
Got IP address: 192.168.1.104
resolved mac address 
Gateway: 192.168.1.10
transfer started ............................... transfer ok, time=1.95s
setting up elf image... OK
jumping to kernel code


OpenWrt kernel loader for AR7XXX/AR9XXX
Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
Decompressing kernel... done!
Starting kernel at 80060000...
```

Reply:
Well, you can just try by increasing LZMA_TEXT_START in here:
https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/image/lzma-loader/Makefile#L11

It's currently chosen to work on 32 MB RAM devices, so if you only build for this device you may increase it and have a look (I don't believe this is the issue however, since the decompressing seems to work fine and the starting is the problem.)

E.g. just try 56 MB:
LZMA_TEXT_START	:= 0x83800000

Reply:
So I did a few tests today moving that value around, and it didn't seem to help.

I **did** get a 5.4 kernel to boot successfully by making a build without the wireless packages, which made a smaller image, and that booted up just fine.

Seems something bad happens when the uncompressed image goes over ~8.5Mb in size.

I'm unsure what though...

Reply:
@NeoRaider You touched the lzma-loader occasionally, do you happen to have an idea what causes the size limitation here?
Boot log (three post above) indicates the decompression is fine, but then nothing happens after the starting line. Playing with LZMA_TEXT_START doesn't have any effect, but I don't really understand why this variable is present twice, once in lzma-loader/Makefile (for compilation) and once in lzma-loader/src/Makefile (for linking) ...

Reply:
Full disclaimer: I know nothing about the mips boot process and I don't know how routerboot handles the boot procedure, I've also not really looked at the lzma-loader code (yet), so I can't really make a suggestion for now. Still:
- To confirm: you got a kernel to boot by reducing the size of the image, not the size of the kernel? That would be significant.
- I'm wondering why we hardcode the kernel partition size in dts, seems like a bad idea? Just looking at the target makefile, it seems that the image is built similarly as the NOR ones, i.e. the kernel and rootfs are embedded in yaffs (via kernel2minor), so I *think* it should be possible to use the minor splitter? (this will probably require adjusting the sysupgrade scripts but it seems to be a good time to do this while we bootstrap mikrotik to ath79: @adrianschmutzler @xback opinons?)

It's still unclear to me why compressing the kernel works around what seems to be an issue with the kernel crossing a memory boundary (AIUI routerboot would fail to load the uncompressed kernel, right?)

I'll take a look at the lzma-loader code and will In any case I don't think this problem should hold this PR back: it's an issue that likely affects or will affects all RB devices and the PR works with the current stable kernel.

Reply:
> In any case I don't think this problem should hold this PR back: it's an issue that likely affects or will affects all RB devices and the PR works with the current stable kernel.

I'm just hoping that if we are able to address this problem properly, we can get rid of the TARGET_INITRAMFS_COMPRESSION_LZMA set for the entire target, as this might have side effects and was discouraged by the other discussion participants during the older issue.

Reply:
> I'm just hoping that if we are able to address this problem properly, we can get rid of the TARGET_INITRAMFS_COMPRESSION_LZMA set for the entire target as well, as this might have side effects and was discouraged by the other discussion participants during the older issue.

That makes sense. It would be nice if someone with an understanding of this architecture could chime in on this issue. I'll dig, but the learning curve is somewhat steep for me :)
Of course, not having a proper spec for routerboot doesn't help 😛

Reply:
Back then, I got some help in IRC (documented in the referenced discussion) as I had to deal with this for the other routerboard PR, and I was only able to patch this in lzma-loader due to the explanations I got there. I fear that for this device, it will be the same, and we will only get "help" if someone of the stakeholders here is pro-actively looking for it.

Reply:
> To confirm: you got a kernel to boot by reducing the size of the image, not the size of the kernel? That would be significant.

I think so? This is a bootable 5.4 build config on top of this PR (i.e. initramfs is compressed), one where I have turned off the wifi packages.

```
$ ./scripts/diffconfig.sh 
CONFIG_TARGET_ath79=y
CONFIG_TARGET_ath79_mikrotik=y
CONFIG_TARGET_ath79_mikrotik_DEVICE_mikrotik_routerboard-493g=y
# CONFIG_DRIVER_11N_SUPPORT is not set
# CONFIG_DRIVER_11W_SUPPORT is not set
CONFIG_LINUX_5_4=y
# CONFIG_PACKAGE_hostapd-common is not set
# CONFIG_PACKAGE_iw is not set
# CONFIG_PACKAGE_kmod-ath is not set
# CONFIG_PACKAGE_kmod-ath9k is not set
# CONFIG_PACKAGE_kmod-cfg80211 is not set
# CONFIG_PACKAGE_kmod-mac80211 is not set
# CONFIG_PACKAGE_wireless-regdb is not set
# CONFIG_PACKAGE_wpad-basic is not set
CONFIG_TESTING_KERNEL=y
```

And binwalk reports the following (and my observation is when the `uncompressed size` goes over circa 8.5Mb it fails to boot.

```
DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             ELF, 32-bit MSB MIPS-I executable, MIPS, version 1 (SYSV)
9532          0x253C          Copyright string: "Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>"
9740          0x260C          LZMA compressed data, properties: 0x6D, dictionary size: 8388608 bytes, uncompressed size: 8178540 bytes
```

Reply:
Please also try this with buildbot config. Building all kmods etc. will increase the kernel by a considerable amount, so it might happen that this is already enough to break it even on 4.19 ...

Reply:
@adrianschmutzler buildbot builds everything but not everything ends up in the image.
@ch6574 I just realized that you're test booting initramfs, that's very significant because then the whole image is loaded in RAM. Have you tried flashing the image and booting from flash?

Reply:
> Please also try this with buildbot config

Is there a setting to do this? (Want to make sure I do it properly).

> Have you tried flashing the image and booting from flash?

Yes, but I've only ever tested the sysupgrade images that first booted via tftp. Do you think I should try flashing a larger build and see if that will boot from the nand?

Reply:
> > Please also try this with buildbot config
> 
> Is there a setting to do this? (Want to make sure I do it properly).

Start from a clean tree. Download https://downloads.openwrt.org/snapshots/targets/ath79/mikrotik/config.buildinfo and rename it to `.config`; then run `make defconfig`.

You can then make menuconfig to disable building SDK and IB to save a bit of time, then run `make`. Wait :)

Reply:
> > Have you tried flashing the image and booting from flash?
> 
> Yes, but I've only ever tested the sysupgrade images that first booted via tftp. Do you think I should try flashing a larger build and see if that will boot from the nand?

If you can flash the default image (i.e. the result of the above configuration) and it boots, that's all we _really_ care about for now :)

Reply:
one step I forgot for buildbot build: run `./script/feeds update` and `script/feeds install -a` before the `make defconfig` step.

Reply:
> buildbot builds everything but not everything ends up in the image.

Yes, but kmods will enable additional symbols that will increase kernel size even if the kmods are not added themselves. I had a case on kirkwood recently where this increased the kernel by something between 40 and 100 kB ...

Reply:
one can also just manually select buildbot settings in the build options menu of make menuconfig.

Reply:
make defconfig doesn't select buildbot config for me.

Just run make menuconfig, select your target/subtarget, and then check "Set build defaults for automatic builds" in "Global build settings" menu.

Disabling SDK and imagebuilder afterwards might help indeed.

Reply:
CONFIG_BUILDBOT is enabled in the config.buildinfo seed, so all is well :)

Reply:
Just did a quick test using this

```
$ ./scripts/diffconfig.sh 
CONFIG_TARGET_ath79=y
CONFIG_TARGET_ath79_mikrotik=y
CONFIG_TARGET_MULTI_PROFILE=y
CONFIG_TARGET_DEVICE_ath79_mikrotik_DEVICE_mikrotik_routerboard-493g=y
CONFIG_TARGET_DEVICE_PACKAGES_ath79_mikrotik_DEVICE_mikrotik_routerboard-493g=""
CONFIG_TARGET_DEVICE_ath79_mikrotik_DEVICE_mikrotik_routerboard-922uags-5hpacd=y
CONFIG_TARGET_DEVICE_PACKAGES_ath79_mikrotik_DEVICE_mikrotik_routerboard-922uags-5hpacd=""
CONFIG_TARGET_DEVICE_ath79_mikrotik_DEVICE_mikrotik_routerboard-wap-g-5hact2hnd=y
CONFIG_TARGET_DEVICE_PACKAGES_ath79_mikrotik_DEVICE_mikrotik_routerboard-wap-g-5hact2hnd=""
CONFIG_ALL_KMODS=y
CONFIG_ALL_NONSHARED=y
CONFIG_DEVEL=y
CONFIG_TARGET_PER_DEVICE_ROOTFS=y
CONFIG_AUTOREMOVE=y
CONFIG_BUILDBOT=y
CONFIG_COLLECT_KERNEL_DEBUG=y
CONFIG_JSON_OVERVIEW_IMAGE_INFO=y
CONFIG_KERNEL_BUILD_DOMAIN="buildhost"
CONFIG_KERNEL_BUILD_USER="builder"
CONFIG_TARGET_ALL_PROFILES=y
# CONFIG_IB is not set
# CONFIG_SDK is not set
```

1. The initramfs image fails to tftp boot (as previously observed, and I see `uncompressed size: 8620268 bytes`)
2. The sysupgrade **does** boot up.

So for step 2 above I booted the device with an older tftp image, and then flashed this build using `sysupgrade -n`, then let it boot from nand and all came up!

I'll re-build this with 5.4 kernel and try flashing that too.

Reply:
@ch6574 thanks this is _very_ helpful. If you can provide a serial boot log (as verbose as possible, I don't know if it's possible to increase routerboot's verbosity at boot) that would also help a lot narrowing down where the problem lives: I have some ideas but I'm only speculating for now.

Reply:
@ch6574 in fact there's another thing you could try to narrow this down: in your target, override KERNEL_INITRAMFS like so: `KERNEL_INITRAMFS := kernel-bin | append-dtb` (i.e. remove the lzma compression and the lzma loader). I'm curious if that would boot?

Reply:
I'm not entirely sure what issue you're hitting (the default LZMA_TEXT_START 0x81800000 and LOADADDR 0x80060000 should be >23MB apart), but it's been a while since I worked on the LZMA loader.

I didn't read the whole PR discussion - why is it necessary to make the initramfs image fit into the flash? Is this required for initial OpenWrt installation?

For most devices, we provide initramfs images for debugging only (which is why they aren't built by default) - they are meant to be loaded via TFTP, to be started directly from RAM without flashing.

Reply:
@NeoRaider as you aptly point out, it makes no sense to apply lzma-loader for initramfs. That's why I want to know if a non-lzma-loader initramfs can boot normally.

Reply:
Just to update, I can flash any size image and it will boot (4.19, 5.4) with everything built as per buildbot config. This is good.

The only thing not working in initramfs images >8.5Mb, and turning on LZMA_COMPRESSION would get a 4.19 image in under this size, but not a 5.4 one.

Reply:
@ch6574 thanks for the feedback, this is helpful. I think I have a hunch what's going on. Can you try building and booting an initramfs as described above? Thanks again much for your time and patience!

Reply:
@f00b4r0

> I don't know if it's possible to increase routerboot's verbosity at boot

It doesn't seem to have anything else to add.

> I want to know if a non-lzma-loader initramfs can boot normally.

I can try something over the weekend on this front.

Reply:
Thanks. If my theory is correct, it will boot :)

Reply:
Actually I just tried `KERNEL_INITRAMFS := kernel-bin | append-dtb`

But it gives: `setting up elf image... not an elf header`

So then tried `KERNEL_INITRAMFS := kernel-bin | append-dtb | loader-kernel`

But that gives: `setting up elf image... kernel out of range`

Reply:
My bad, this needs a little more work. I'll get back to you tomorrow, too late for me now. Thanks again